0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+
+## Economics API
+Accounts own certain balance; and each transaction and each receipt have certain amount of balance attached to them.
+During the contract execution, the contract has access to the following `u128` values:
+* `account_balance` -- the balance attached to the given account;
+* `attached_balance` -- the balance that was attached to the receipt:
+  * If this receipt is a direct result of a transaction that does function call then the balance attached to the
+    transaction is reduced by the transaction fee amount and attached to the receipt;
+  * If this receipt is a result of cross-contract call then the code has to explicitly specify how much is attached to
+    the cross-contract call.
+* `remaining_balance` -- the balance that the contract code uses for:
+  * Calling `deposit`;
+  * Wasm execution fees, like cost of operations, growth of stack, etc.
+  * Fees for calling host functions, including fees for creating promises, **including balance attached to these promises.**
+  
+  This balance can be refilled by calling `withdraw` function.
+* `burnt_balance` -- the balance that the contract already used for:
+  * Wasm exeuction fees;
+  * Fees for calling host functions, including fees for creating promises, **excluding balance attached to these promises.**
+  
+  This balance cannot be recovered, if the contract execution fails.
+  
+When contract execution fails the following happens to the balances:
+* First we compute `deficit = all_withdrawals - all_deposits`;
+* If `deficit > 0` (we have withdrawn from account more than we deposited to it):
+  * If `burnt_balance <= attached_balance` then we deposit `deficit` back to the `account_balance` and issue
+    a refund of `attached_balance - burnt_balance` if this value is not zero.
+  * If `attached_balance < burnt_balance <= attached_balance + deficit` then we deposit
+  `attached_balance + deficit - burnt_balance` back to the `account_balance` and do not issue a refund;
+* If `deficit <= 0` (we have deposited to the account more than what we have withdrawn) we withdraw `abs(deficit)`
+  from the `account_balance` and issue a refund of `attached_balance - burnt_balance`.
+  
+Note, for both cases above, `burnt_balanace` is never greater than `attached_balance + deficit` at any point in the program. This
+might require a formal proof later.
+
+---
+The following spec is the same for all functions:
+```
+account_balance(balance_ptr: u64)
+attached_balance(balance_ptr: u64)
+remaining_balance(balance_ptr: u64)
+burnt_balance(balance_ptr: u64)
+```
+ -- writes the value into the `u128` variable pointed by `balance_ptr`.
+
+###### Panics
+* If `balance_ptr + 16` points outside the memory of the guest with `MemoryAccessViolation`;
+
+###### Current bugs
+* Use a different name;
+* Unclear how they behave with the callbacks.
+
+---
+```rust
+deposit(min_amount_ptr: u64, max_amount_ptr: u64, result_ptr: u64)
```

Comment:
We'll have to revisit deposit and withdrawals if we split `fee` and `amount` balances

Reply:
Done.